google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java

```diff
@@ -144,8 +146,31 @@ public static CloudStorageFileSystem forBucket(String bucket, CloudStorageConfig
     checkArgument(
         !bucket.startsWith(URI_SCHEME + ":"), "Bucket name must not have schema: %s", bucket);
     checkNotNull(config);
-    return new CloudStorageFileSystem(
-        new CloudStorageFileSystemProvider(config.userProject()), bucket, config);
+    return new CloudStorageFileSystem(getCloudStorageFileSystemProvider(config), bucket, config);
+  }
+
+  /**
+   * Creates new file system provider instance for {@code CloudStorageFileSystemProvider}.
+   *
+   * @param config CloudStorageConfiguration to initialize {@code CloudStorageFileSystemProvider}
+   *     object
+   * @param storageOptions
+   * @return {@code CloudStorageFileSystemProvider} object
+   */
+  private static CloudStorageFileSystemProvider getCloudStorageFileSystemProvider(
+      CloudStorageConfiguration config, StorageOptions... storageOptions) {
+    CloudStorageFileSystemProvider cloudStorageFileSystemProvider =
+        (storageOptions.length >= 1)
+            ? new CloudStorageFileSystemProvider(config.userProject(), storageOptions[0])
+            : new CloudStorageFileSystemProvider(config.userProject());
```

Comment:
you need to create new provider only if the map does not contain one

Reply:
i have to create provider instance every time b'z some time conifiguration remine same but  provider may be different due to storageoption paramer so that's why i have create it for comparision.
    
    config object same=>provider1(storageopton with a crentialfile) 
     config object same=>provider2 (storageopton with b crentialfile)
     
b'z here configuration is same only upgradation in provider was required.

Reply:
If you create a new object each time, no need to have a map.
Would it make sense in the implementation of  `public static Path copy(Path source, Path target, CopyOption... options) throws IOException` replace `if (provider(target) == provider) {` with `if (provider(target).equals(provider)) {`